File: euler25.py
# Каков порядковый номер первого члена последовательности
# Фибоначчи, содержащего 1000 цифр?

from datetime import datetime


def fibonacci(f1=0, f2=1):
    """Итерируемая функция вычисления следующего числа Фибоначчи"""
    while True:
        yield f1
        f1, f2 = f2, f1 + f2


def euler25():
    n = 1_000

    # 1 вариант
    start_time = datetime.now()
    fb_counter = 0
    iterator = fibonacci()
    fb_num = next(iterator)
    while len(str(fb_num)) < n:
        fb_counter += 1
        fb_num = next(iterator)
    print(fb_counter)
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    fb_counter = 0
    for i in fibonacci():
        if len(str(i)) == n: break
        fb_counter += 1
    print(fb_counter)
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler25()

File: go/euler25.go
// euler25 находит порядковый номер первого числа Фибоначчи с 1000 цифрами.
//
// Решение задачи Эйлера №25.
package main

import (
	"fmt"
	"math/big"
	"time"
)

func euler25() {
	n := 1000

	startTime := time.Now()
	fbCounter := 0
	f1 := big.NewInt(0)
	f2 := big.NewInt(1)
	fbNum := big.NewInt(0)
	one := big.NewInt(1)
	ten := big.NewInt(10)
	pow := big.NewInt(int64(n - 1))
	limit := ten.Exp(ten, pow, nil)
	for {
		if fbNum.BitLen() >= n {
			break
		}
		fbCounter++
		fbNum.Add(f1, f2)
		f1, f2 = f2, fbNum
	}
	fmt.Println(fbCounter)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fbCounter = 0
	f1 = big.NewInt(0)
	f2 = big.NewInt(1)
	for {
		if len(f1.String()) == n {
			break
		}
		fbCounter++
		f1.Add(f1, f2)
		f1, f2 = f2, f1
	}
	fmt.Println(fbCounter)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler25()
}
